src/java/org/apache/cassandra/db/monitoring/SaiSlowLog.java

```diff
@@ -0,0 +1,193 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.monitoring;
+
+import java.beans.ConstructorProperties;
+import java.util.Comparator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.tracing.Tracing;
+
+public class SaiSlowLog
+{
+    private static final Logger logger = LoggerFactory.getLogger(SaiSlowLog.class);
+
+    private final AtomicInteger numSlowestQueries = new AtomicInteger();
+    // a structure that keeps the N slowest queries - don't use getter/setter so that it's not exposed over JMX
+    private final SaiSlowestQueriesQueue slowestQueries;
+    // KATE: make this false by default and decide on yaml property, -D flag, or JMX to enable; hook setEnabled() to change
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
+
+    static final Comparator<SlowSaiQuery> SLOW_SAI_QUERY_COMPARATOR = (SlowSaiQuery o1, SlowSaiQuery o2) -> Long.compare(o2.getDuration(), o1.getDuration());
+
+    @VisibleForTesting
+    static SaiSlowLog instance = new SaiSlowLog();
+
+    public SaiSlowLog()
+    {
+        numSlowestQueries.set(DatabaseDescriptor.getSaiSlowLogNumSlowestQueries());
+        slowestQueries = new SaiSlowestQueriesQueue(numSlowestQueries.get());
+    }
+
+    public boolean isEnabled()
+    {
+        return enabled.get();
+    }
+
+    public synchronized void setEnabled(boolean newValue)
+    {
+        boolean oldValue = this.enabled.get();
+
+        if (oldValue != newValue)
+        {
+            this.enabled.set(newValue);
+            logger.info("SAI Slow Log is now {}", newValue ? "enabled" : "disabled");
+        }
+    }
+
+    // KATE below two will be used to set and get the number of slowest queries; make it configurable through property/nodetool?
+    public int getNumSlowestQueries()
+    {
+        return numSlowestQueries.get();
+    }
+
+    public void setNumSlowestQueries(int numSlowestQueries)
+    {
+        int oldValue = this.numSlowestQueries.get();
+        if (oldValue != numSlowestQueries)
+        {
+            if (numSlowestQueries < 1)
+            {
+                throw new IllegalArgumentException("sai_slow_log_num_slowest_queries must be > 0");
+            }
+            this.numSlowestQueries.set(numSlowestQueries);
+            resizeMinMaxHeap();
+            logger.info("sai_slow_log_num_slowest_queries set to {} (was {}).", numSlowestQueries, oldValue);
+        }
+    }
+
+    private void resizeMinMaxHeap()
```

Comment:
No need to make this resizable at runtime

Reply:
I would prefer to not keep code that we won't use.

as said in other comments:
- we can restart the JVM to change the value (and I am not sure that we will often change it)
- we should add new http endpoints to change the configuration and add docs about how to use this, for now we don't have a framework to handle this easily


if we want to keep the code to do the dynamic reconfig I am fine with it, but we cannot left the code there without the full story end-to-end.

Reply:
`if we want to keep the code to do the dynamic reconfig I am fine with it, but we cannot left the code there without the full story end-to-end.`
I agree and this was never suggested. I did not have things exposed yet because this was not finished and I wanted to be sure that is what we want first before exposing config in nodetool, JMX, HTTP endpoint